src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/AwsSdk/AwsSdkPipelineWrapper.cs

```diff
@@ -74,19 +82,67 @@ public AfterWrappedMethodDelegate BeforeWrappedMethod(InstrumentedMethodCall ins
 
             string requestQueueUrl = request.QueueUrl;
             ISegment segment = SqsHelper.GenerateSegment(transaction, instrumentedMethodCall.MethodCall, requestQueueUrl, action);
-            if (insertDistributedTraceHeaders)
+            if (action == MessageBrokerAction.Produce)
             {
-                // This needs to happen at the end
-                if (requestContext.Request == null)
-                    agent.Logger.Finest("AwsSdkPipelineWrapper: requestContext.Request is null, unable to insert distributed trace headers.");
+                if (request.MessageAttributes == null)
+                {
+                    agent.Logger.Debug("AwsSdkPipelineWrapper: requestContext.OriginalRequest.MessageAttributes is null, unable to insert distributed trace headers.");
+                }
                 else
                 {
-                    dynamic webRequest = requestContext.Request;
-                    SqsHelper.InsertDistributedTraceHeaders(transaction, webRequest);
+                    SqsHelper.InsertDistributedTraceHeaders(transaction, request);
                 }
             }
 
-            return Delegates.GetDelegateFor(segment);
+            // modify the request to ask for DT headers in the response message attributes
+            if (action == MessageBrokerAction.Consume)
+            {
+                if (request.MessageAttributeNames == null)
+                    request.MessageAttributeNames = new List<string>();
+
+                request.MessageAttributeNames.Add(NEWRELIC_TRACE_HEADER);
+                request.MessageAttributeNames.Add(W3C_TRACESTATE_HEADER);
+                request.MessageAttributeNames.Add(W3C_TRACEPARENT_HEADER);
```

Comment:
We might want to add an experimental extensions API that allows us to get the actual list of configured headers, because we won't always have 3. Sometimes it might just be 1 or 2. Or CAT could be enabled, and none of these 3 will be used.

Reply:
Good idea. I'll look into that in a separate PR.